Improve config, send transaction ID via HTTP header

lipusal opened the pull request.

Commit:
First attempt

nicoayala commented on lib/log_weasel/middleware.rb:
```diff
@@ -1,16 +1,34 @@
 class LogWeasel::Middleware
   KEY_HEADER = 'X_LOGWEASEL_KEY'
+  TRANSACTION_ID_HEADER = 'X_LOGWEASEL_TRANSACTION_ID'
```
Creo que seria mejor mantener el nombre del header genérico: `X_TRANSACTION_ID` (estaria bueno que sea configurable, ver [LogWeasel.config](https://github.com/carbonfive/log_weasel/blob/master/lib/log_weasel.rb))
Desde el punto de vista de un cliente que le pega a `redo-api` enviando el header por ejemplo, porque deberia saber que usamos `log_weasel`?

nicoayala commented on lib/log_weasel/middleware.rb:
```diff
@@ -1,16 +1,34 @@
-    LogWeasel::Transaction.create key
-    @app.call(env)
+    # Get or create transaction ID
+    transaction_id = env.fetch("HTTP_#{TRANSACTION_ID_HEADER}", nil)
```
Esto lo podemos resumir a:
```ruby
transaction_id = env.fetch("HTTP_#{TRANSACTION_ID_HEADER}", LogWeasel::Transaction.create(key))
LogWeasel::Transaction.id = transaction_id
```

nicoayala reviewed (changes requested):
Lo veo bien! Van unos comments para unos cambios.

Commit:
Use standard configuration classes

Commit:
Set default config

Commit:
Only add middleware if enabled, delete unused code

Commit:
Use config for generating transaction IDs

Commit:
Refactor middleware

- Use new config
- Stop using key
- Don't attach transaction ID if missing

lipusal commented:
@nicoayala Gracias por tus comments, me di la libertad de ir un poco más allá con la configuración

- El header es configurable
- Toda la gema se puede prender y apagar (algo así tiene `lograge`)
- Se puede pasar un generador de IDs custom, por default usa el que ya tiene
- Se puede configurar que si no hay transaction ID que no lo genere

lipusal commented:
También saqué eso de que prefije los IDs con una clave, no sé para que estaba. Creo que era para si un mismo thread corría más de una aplicación con `log_weasel` para evitar clashes? Creo que no hace falta, si te parece lo vuelvo a poner

Commit:
Use Hashie::Mash for configuration, fix key generation

- ActiveSupport options was not taking default options, and was old
- Add Hashie dependency and use it for configuration
- Move default ID generator to config
- Fix call to ID generator

Commit:
Fix transaction specs

- Remove key specs
- Use newer rspec notation

Commit:
Fix log weasel specs

- Remove key spec
- Add new specs

nicoayala reviewed (changes requested):
Lo veo casi listo, van algunos comments:
* Sacaría la dependencia con `hashie` y usaria la clase `Config` como estaba antes.
* No me queda claro si es necesario el atributo `key` que mencionás, parecería que no, pero miremos algun log generado sin `key` para entender un poco mejor.

Commit:
Merge branch 'master' into feature/persist-ids

# Conflicts:
#	lib/log_weasel.rb
#	lib/log_weasel/middleware.rb
#	lib/log_weasel/railtie.rb
#	lib/log_weasel/transaction.rb
#	log_weasel.gemspec
#	spec/log_weasel/log_weasel_spec.rb
#	spec/log_weasel/transaction_spec.rb

Commit:
Missing Rubocop fixes

Commit:
Remove hashie dependency

- Use PORO (is that a word?) to store config again

lipusal commented:
@nicoayala Parece que se usaba como [prefijo de las transaction IDs](https://github.com/restorando/log_weasel/blob/b26da8d4c33dc7071b2336b6868252ebca52b403/lib/log_weasel/transaction.rb#L11), creo que no hace falta. Se utilizaba porque creo que un mismo thread podía correr varias aplicaciones distintas con log weasel, ahora ya no. De todas maneras sí, vamos a hacer pruebas a ver cómo cambia.
Btw yay Travis 🎉

Commit:
Extract code to conditionally generate ID to method in Transaction

- Add LogWeasel::Transaction#set_or_create
- Use new method in middleware

Commit:
Fix rubocop

lipusal commented on lib/log_weasel/transaction.rb:
```diff
@@ -19,9 +18,23 @@ def self.id=(id)
+    # ID will be generated.
+    # @return [Object] The set or generated transaction ID. `nil` if neither set not generated.
+    # rubocop:disable Style/AccessorMethodName
+    def self.set_or_create(id = nil)
```
Quizá hay un mejor nombre para este método? A Rubocop no le gusta el prefijo `set_`

lipusal commented:
This has been showing up in my PR feed for years now so I'm gonna close it =P
